### PR TITLE
feat: Chat search

### PR DESCRIPTION
**Pull request by drminh2807**
# Depend on
- https://github.com/Sorunome/matrix_link_text/pull/4

# Issue
- https://www.notion.so/As-a-user-I-want-to-perform-search-inside-chat-89b4e3fd3cbb481cb7633bf2b81855af

# Demo
## Mobile scroll

https://github.com/linagora/twake-on-matrix/assets/12546908/25d77f77-1b6f-428b-943b-6901afffc17d

## Mobile highlight

https://github.com/linagora/twake-on-matrix/assets/12546908/3b8dc568-fb22-4622-905c-fdd86cd1860c

## Mobile navigate from chat detail

https://github.com/linagora/twake-on-matrix/assets/12546908/c03c5033-8193-4442-8bfb-077d4c7df8f2

## Web searching

https://github.com/linagora/twake-on-matrix/assets/12546908/c6f5bce8-e7b4-461b-9cf6-ce81020e292b

## Web: Navigate from chat detail

https://github.com/linagora/twake-on-matrix/assets/12546908/c0fdec13-f985-4e29-97fa-1e6e7fbfac81

### CODE REVIEW

**Comment by hoangdat** [comment]
Please write an ADR for this solution.

File: lib/utils/string_extension.dart [hunk 1]
```diff
@@ -146,4 +147,74 @@ extension StringCasingExtension on String {
         timeline.events.indexOf(secondEvent) >
             timeline.events.indexOf(firstEvent);
   }
+
+  bool containsWord(String word) {
+    final regExp = RegExp(
+      "\\b(?:$word)\\b",
```
Comment (hoangdat) [inline lib/utils/string_extension.dart:153]: what is it? how to make it more clear for the blind of regex?

File: lib/utils/string_extension.dart [hunk 2]
```diff
@@ -146,4 +147,74 @@ extension StringCasingExtension on String {
...
+      caseSensitive: false,
+    );
+    return regExp.hasMatch(this);
+  }
+
+  String htmlHighlightText(targetText) {
+    final regex = RegExp(
+      '(<[^>]*>)|(${RegExp.escape(targetText)})',
```
Comment (hoangdat) [inline lib/utils/string_extension.dart:161]: idem

**Comment by sherlockvn** [comment]
> # Depend on
> * [feat: Add text span builder Sorunome/matrix_link_text#4](https://github.com/Sorunome/matrix_link_text/pull/4)
> 
> # Issue
> * https://www.notion.so/As-a-user-I-want-to-perform-search-inside-chat-89b4e3fd3cbb481cb7633bf2b81855af
> 
> # Demo
> ## Mobile scroll
>  Screen.Recording.2023-09-08.at.16.09.03.mov 
> ## Mobile highlight
>  Screen.Recording.2023-09-08.at.16.14.14.mov 
> ## Web
>  Screen.Recording.2023-09-08.at.16.18.30.mov

The interval when jump to message should be small or directly to, otherwise the application may be down performance a lot when jump between message

File: lib/config/app_config.dart [hunk 3]
```diff
@@ -69,6 +69,8 @@ abstract class AppConfig {
   static const double borderRadius = 20.0;
   static const double columnWidth = 360.0;
   static const int fetchContactsLimit = 20;
+  static const int chatRoomSearchKeywordMin = 2;
+  static const bool chatRoomSearchWordStrategy = false;
```
Comment (sherlockvn) [inline lib/config/app_config.dart:73]: strategy = false ? @@

File: lib/domain/usecase/room/chat_room_search_interactor.dart [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            ChatRoomSearchSuccess(keyword: keyword, eventIndex: index),
+          );
+          return;
+        }
+        yield Left(ChatRoomSearchNoResult());
+        return;
+      }
+      for (var i = 0; i < maxRequestHistory; i++) {
+        final index =
+            timeline.events.sublist(currentIndex + 1).indexWhere(searchFunc);
```
Comment (sherlockvn) [inline lib/domain/usecase/room/chat_room_search_interactor.dart:46]: this function don't depend on `i`, why put it inside the for loop

File: lib/pages/chat/chat_room_search_mixin.dart [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+import 'package:debounce_throttle/debounce_throttle.dart';
+import 'package:fluttertoast/fluttertoast.dart';
+import 'package:get_it/get_it.dart';
+import 'package:matrix/matrix.dart';
+import 'package:flutter_gen/gen_l10n/l10n.dart';
+
+mixin ChatRoomSearchMixin {
+  static const int _maxRequestHistory = 10;
+
+  final isSearching = ValueNotifier(false);
```
Comment (sherlockvn) [inline lib/pages/chat/chat_room_search_mixin.dart:17]: `isSearchinNotifier`

File: lib/pages/chat/chat_room_search_mixin.dart [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+  final searchTextController = TextEditingController();
+  final searchStatus =
+      ValueNotifier<Either<Failure, Success>>(Right(SearchInitial()));
+  final canGoUp = ValueNotifier(false);
+  final canGoDown = ValueNotifier(false);
+
+  final _searchInteractor = GetIt.I.get<ChatRoomSearchInteractor>();
```
Comment (sherlockvn) [inline lib/pages/chat/chat_room_search_mixin.dart:24]: `getIt.get<....>`

File: lib/pages/chat/chat_room_search_mixin.dart [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+  int _historyCount = 100;
+  Timeline? Function()? _getTimeline;
+  void Function(int)? _scrollToIndex;
+
+  static const _debouncerDuration = Duration(milliseconds: 300);
+  final _debouncer = Debouncer(
```
Comment (sherlockvn) [inline lib/pages/chat/chat_room_search_mixin.dart:31]: cancel() it when being dispose

File: lib/config/app_config.dart (hunk 3 above)
Comment (drminh2807) [reply lib/config/app_config.dart:73]: disable for now, we will support it if required

File: lib/pages/chat/chat_room_search_mixin.dart (hunk 5 above)
Comment (drminh2807) [reply lib/pages/chat/chat_room_search_mixin.dart:17]: done

File: lib/pages/chat/chat_room_search_mixin.dart (hunk 6 above)
Comment (drminh2807) [reply lib/pages/chat/chat_room_search_mixin.dart]: done

File: lib/pages/chat/chat_room_search_mixin.dart (hunk 7 above)
Comment (drminh2807) [reply lib/pages/chat/chat_room_search_mixin.dart:31]: done

File: lib/domain/usecase/room/chat_room_search_interactor.dart [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+import 'package:fluffychat/domain/app_state/room/chat_room_search_state.dart';
+import 'package:fluffychat/utils/string_extension.dart';
+import 'package:matrix/matrix.dart';
+
+class ChatRoomSearchInteractor {
+  Stream<Either<Failure, Success>> execute({
+    required Timeline timeline,
+    required String keyword,
+    required int currentIndex,
+    required bool isGoDown,
```
Comment (hoangdat) [inline lib/domain/usecase/room/chat_room_search_interactor.dart:14]: why not use the same as Matrix SDK: `Direction.b` or `Direction.f`

File: lib/domain/usecase/room/chat_room_search_interactor.dart [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+import 'package:fluffychat/config/app_config.dart';
+import 'package:fluffychat/domain/app_state/room/chat_room_search_state.dart';
+import 'package:fluffychat/utils/string_extension.dart';
+import 'package:matrix/matrix.dart';
+
+class ChatRoomSearchInteractor {
+  Stream<Either<Failure, Success>> execute({
+    required Timeline timeline,
+    required String keyword,
+    required int currentIndex,
```
Comment (hoangdat) [inline lib/domain/usecase/room/chat_room_search_interactor.dart:13]: what is it? Index of what?

File: lib/domain/usecase/room/chat_room_search_interactor.dart (hunk 4 above)
Comment (hoangdat) [reply lib/domain/usecase/room/chat_room_search_interactor.dart:46]: srr, but not easy to understand it.

Comment (hoangdat) [reply lib/domain/usecase/room/chat_room_search_interactor.dart:46]: Can you describe more details about `maxRequestHistory`?

File: lib/domain/usecase/room/chat_room_search_interactor.dart [hunk 10]
```diff
@@ -0,0 +1,66 @@
+import 'package:dartz/dartz.dart';
+import 'package:fluffychat/app_state/failure.dart';
+import 'package:fluffychat/app_state/success.dart';
+import 'package:fluffychat/config/app_config.dart';
+import 'package:fluffychat/domain/app_state/room/chat_room_search_state.dart';
+import 'package:fluffychat/utils/string_extension.dart';
+import 'package:matrix/matrix.dart';
+
+class ChatRoomSearchInteractor {
+  Stream<Either<Failure, Success>> execute({
```
Comment (hoangdat) [inline lib/domain/usecase/room/chat_room_search_interactor.dart:10]: can you write the unit-tests for this method?

**Comment by hoangdat** [comment]
Hanlde onTap on Search button of Chat Details (GroupChat)

File: lib/utils/string_extension.dart (hunk 2 above)
Comment (drminh2807) [reply lib/utils/string_extension.dart:161]: done

File: lib/utils/string_extension.dart (hunk 1 above)
Comment (drminh2807) [reply lib/utils/string_extension.dart:153]: it is easily to understand if place next to `contains` exist function
if need more clear, I think we should add docs comment for that function
Screenshot 2023-09-12 at 11 25 55

File: test/string_extension_test.dart [hunk 11]
```diff
@@ -190,4 +190,81 @@ void main() {
...
+  });
+
+  test('Does not contain word - Case sensitive', () {
+    expect(sampleString.containsWord('Foo'), false);
+    expect(sampleString.containsWord('BAR'), false);
+  });
+
+  test('Empty string', () {
+    expect(''.containsWord('word'), false);
+  });
```
[the remark this answers is not in the document]
Comment (drminh2807) [reply test/string_extension_test.dart]: done

File: test/string_extension_test.dart [hunk 12]
```diff
@@ -190,4 +190,81 @@ void main() {
...
+  test('Highlight text in HTML', () {
```
[the remark this answers is not in the document]
Comment (drminh2807) [reply test/string_extension_test.dart]: done

File: pubspec.yaml [hunk 13]
```diff
@@ -253,6 +253,10 @@ dependency_overrides:
       url: https://github.com/chandrabezzo/wakelock.git
       ref: main
       path: wakelock_windows/
+  matrix_link_text:
```
[the remark this answers is not in the document]
Comment (drminh2807) [reply pubspec.yaml:258]: done

File: lib/domain/usecase/room/chat_room_search_interactor.dart (hunk 9 above)
Comment (drminh2807) [reply lib/domain/usecase/room/chat_room_search_interactor.dart]: updated to currentEventIndex

Comment (drminh2807) [reply lib/domain/usecase/room/chat_room_search_interactor.dart]: you can read the ADR to make clear about search logic
- Employ Dart for finding the index of the first matching message.
- When the user presses next/previous, search for the index of the next matching message.

**Review by hoangdat — APPROVED** [review]

File: lib/pages/chat/chat_room_search_mixin.dart [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+      keyword: keyword,
+      currentEventIndex: -1,
+      direction: Direction.b,
+      limitPerRequest: _historyCount,
+      maxRequest: _maxRequestHistory,
+    )
+        .listen(
+      (event) {
+        Logs().v(
+          "ChatRoomSearchMixin::search $event",
```
Comment (imGok) [inline lib/pages/chat/chat_room_search_mixin.dart:78]: IMO same log in 2 different functions is not good for debug
```suggestion
          "ChatRoomSearchMixin::_handleSearch $event",
```

File: lib/pages/chat/chat_room_search_mixin.dart [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+      keyword: lastSuccess.keyword,
+      currentEventIndex: lastSuccess.eventIndex,
+      direction: direction,
+      limitPerRequest: _historyCount,
+      maxRequest: _maxRequestHistory,
+    )
+        .listen(
+      (event) {
+        Logs().v(
+          "ChatRoomSearchMixin::search $event",
```
Comment (imGok) [inline lib/pages/chat/chat_room_search_mixin.dart:126]:
```suggestion
          "ChatRoomSearchMixin::_goSearchResult $event",
```

File: lib/pages/chat/chat_view_style.dart [hunk 16]
```diff
@@ -14,4 +14,24 @@ class ChatViewStyle {
...
         start: responsive.isMobile(context) ? 0 : 16,
       );
+
+  static BoxDecoration searchBottomViewDecoration(BuildContext context) =>
+      BoxDecoration(
+        color: Theme.of(context).colorScheme.surface,
+      );
+
+  static EdgeInsetsDirectional get noSearchResultPadding =>
+      const EdgeInsetsDirectional.all(8);
```
Comment (imGok) [inline lib/pages/chat/chat_view_style.dart:24]: you said const was better for the performance, is it as efficient as this ? 

```suggestion
  static const EdgeInsetsDirectional noSearchResultPadding = EdgeInsetsDirectional.all(8);
```

File: lib/widgets/highlight_text.dart [hunk 17]
```diff
@@ -16,51 +17,20 @@ class HighlightText extends StatelessWidget {
 
   @override
   Widget build(BuildContext context) {
-    final spans = buildTextSpans();
+    final acutalHighlightStyle = highlightStyle ??
```
Comment (imGok) [inline lib/widgets/highlight_text.dart:20]: typo
```suggestion
    final actualHighlightStyle = highlightStyle ??

```

**Review by sherlockvn — APPROVED** [review]

File: lib/pages/chat/chat_room_search_mixin.dart (hunk 14 above)
Comment (drminh2807) [reply lib/pages/chat/chat_room_search_mixin.dart]: done

File: lib/pages/chat/chat_room_search_mixin.dart (hunk 15 above)
Comment (drminh2807) [reply lib/pages/chat/chat_room_search_mixin.dart]: done

File: lib/pages/chat/chat_view_style.dart (hunk 16 above)
Comment (drminh2807) [reply lib/pages/chat/chat_view_style.dart]: done

File: lib/widgets/highlight_text.dart (hunk 17 above)
Comment (drminh2807) [reply lib/widgets/highlight_text.dart]: done

**Review by nqhhdev — APPROVED** [review]

**Review by hoangdat — APPROVED** [review]

**Pull request merged**